feat: reworks data planes list view

kleinfreund opened the pull request:
## Changes

Reworks the data plane list view. The bottom portion of the view detailing the currently selected data plane was removed. As replacement, a condensed data plane summary of the selected data plane was added in the form of a fixed sidebar to the right of the data planes table. Each data plane summary includes a link to a newly added data plane detail view. The data planes table was tweaked to be more space-efficient. Only a few specific tags (kuma.io/service, kuma.io/protocol, and kuma.io/zone) are now represented in the table as separate columns. The sidebar lists all tags.

Adds a column selector to the data plane list view which allows controlling which columns are visible in the table. The selection is stored in the user browser’s local storage (i.e. its device-bound). Eventually, this selector will be added to all data tables in the UI.

Adds a data plane detail view and associated routes. It currently shows the same tabular overview that used to be shown below the data planes table in the list view.

Moves all data plane specific files into `src/data-planes`. This is motivated by wanting to structure the code base by the information hierarchy in the UI.

Creates LoadingBlock, ErrorBlock, and EmptyBlock components for simplifying the use of commonly found components (i.e. the empty states shown when loading or when content is empty or couldn’t be loaded).

Restructures the way the top level elements in the UI (i.e. the main sidebar, main content area, top bar, etc.) are laid out. This was necessary to gain a better control over the layout possibilities for the changes made to the data planes list view. As a nice side effect, a lot of unused or unnecessary code could be removed (e.g. controlling the hover state of the main navigation via CSS instead of JS).

Updates Kongponents to the latest version. This is necessary because some bug fixes made in between are necessary for the changes to the data plane views to work properly. Unfortunately, this made it necessary to regenerate all snapshots which accounts for the biggest number of files changed in this commit.

## Testing

### With mock data

```sh
yarn run dev
```

### With fake data planes

Terminal window 1:

```sh
cd kuma
KUMA_DP_SERVER_AUTH_TYPE=none make run/universal/memory
```

Terminal window 2:

```sh
cd kuma
make run/xds-client
```

Terminal window 3:

```sh
cd kuma-gui
yarn run dev:with-api
```

## Screenshots

<details>
<summary>Before</summary>

![Screenshot from 2022-09-14 10-39-49](https://user-images.githubusercontent.com/5774638/190105443-31f600d4-948c-4745-9450-7ccc694c4ab0.png)

</details>

<details>
<summary>After</summary>

![image](https://user-images.githubusercontent.com/5774638/190185787-49b86b2b-bfbc-4c67-9caf-1dfc938653b0.png)

</details>

Resolves #382.

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

kleinfreund reviewed (commented):
**Question**: Should data planes in the list view be sorted by associated service first?

lahabana commented:
In dataplane detail view. The link in the breadcrumb to "Data plane" is broken.

lahabana commented:
Should we get rid of "Entity Overview for backend" in the sidebar view I'm not sure it brings much

lahabana commented:
In the details view. When clicking yaml can we just call the `dataplane` endpoint to just retrieve the actual dataplane instead of the dataplane overview?

lahabana commented:
Ok how about a follow up fix to the detail view (I think YAML should just go in `general` and tabs should be reflected in the url)

lahabana commented on src/data-planes/views/DataPlaneListView.vue:
```diff
@@ -0,0 +1,654 @@
+    {
+      label: 'Gateway',
+      type: 'Gateway / BUILTIN',
+    },
```
Are we missing provided gateways here?

kleinfreund replied:
I’ve added an entry for `Gateway / PROVIDED`. I *assume* this is how the type would be constructed in the UI though I’m not sure because there is no mock data that features this type. I suppose the `gateway.type` field would be set to `'PROVIDED'` for those?

kleinfreund replied:
A few questions:

- I’ve seen some test examples in Kuma that suggest that `gateway.type` isn’t typically all uppercase. However, this is the case in our mock data. If it’s lowercase, I’m going to change the mock data.
- “Gateway / BUILTIN” looks strange to me. I suggest a slightly different format for the UI: “Gateway (builtin)” (and “Gateway (provided)”). What do you think?

lahabana replied:
These are the 2 different ways to run gateways: https://kuma.io/docs/1.8.x/explore/gateway/

Going to try to find enough time to start the 2 gateways and get you the API output. FTM I think it's fine to merge as is and open a follow up issue if it's addressed before 2.0

lahabana commented on src/data-planes/views/DataPlaneListView.vue:
```diff
@@ -0,0 +1,654 @@
+      /**
+       * Handle our tag collections based on the dataplane type.
+       */
+      const tags = dpTags(dataplane).filter((tag) => IMPORTANT_DATA_PLANE_TAG_LABELS.includes(tag.label))
```
What's the of extracting this as a constant when the 3 entries in your constant are used just after? Seems like this would be better inlined.

kleinfreund replied:
I’m not sure I’m following.

Do you mean to suggest to look up service, protocol, and zone on `dataplane` directly rather than constructing the tag list and then finding them again?

lahabana replied:
I'm just saying the global var seems not needed here. It's just asking the reader of the code to context switch when it's only used here and the 3 following lines would need to evolve accodingly to this global var.

kleinfreund replied:
Ah, now I understand. Done.

lahabana commented on src/dataplane.ts:
```diff
@@ -1,70 +1,88 @@
+]
+
+/**
+ * Takes a data plane and constructs the list of tags. It flattens common tags so we don't display them twice.
```
This seems to contradict your example

kleinfreund replied:
The way this utility was implemented considers uniqueness of tags as label-value pairs meaning that a network inbound tag `kuma.io/service: frontend` and a network outbound tag `kuma.io/service: backend` are distinct.

lahabana replied:
Ok might be worth clarifying the comment.

kleinfreund replied:
I’ve added a note to it.

lahabana reviewed (approved):
Just some minor comments

kleinfreund commented:
@lahabana 

> In dataplane detail view. The link in the breadcrumb to "Data plane" is broken.

What’s broken exactly? I fixed one issue where the URL still contained a `name` parameter when navigating away from the list view, but that seems unrelated.

> Should we get rid of "Entity Overview for backend" in the sidebar view I'm not sure it brings much

I agree; I don’t care much for it and in most places it’s really just noisy. I removed it.

> In the details view. When clicking yaml can we just call the `dataplane` endpoint to just retrieve the actual dataplane instead of the dataplane overview?

I’ve reworked the detail view to generally use the `Dataplane` entity which also addresses this point.

---

What’s your opinion on https://github.com/kumahq/kuma-gui/pull/384#pullrequestreview-1113809561?

lahabana commented:
> Question: Should data planes in the list view be sorted by associated service first?

Good question I think you should keep as is for now. It would be nice to have it sorted but I'd do this separately

lahabana commented:
What I meant about breadcrumbs is here:

<img width="524" alt="image" src="https://user-images.githubusercontent.com/702355/191488096-ff222bfe-ceb9-484a-9e6a-1468e9932942.png">


I'd expect that clicking on `Data plane` would redirect to the list view.

kleinfreund commented:
@lahabana

> I'd expect that clicking on `Data plane` would redirect to the list view.

This is now fixed.
